feat(explore): export csv data pivoted for Pivot Table [ID-9]

kgabryje opened the pull request:
### SUMMARY
Apply pivot table post processing to data downloaded as CSV in Pivot Table V2 chart. The CSV data will have pivoted format (see the difference on screenshots).

When on Pivot Table chart, the .CSV button is replaced by a dropdown containing 2 options - Original and Pivoted. Clicking the latter triggers download of pivoted data.
On all other charts, the CSV button works just like before.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Example pivot table:

![image](https://user-images.githubusercontent.com/15073128/142918197-0363b4d6-e04f-43a0-847c-a0ab54ab5b46.png)

Before:
![image](https://user-images.githubusercontent.com/15073128/142918482-cb176400-aafa-41f0-8060-0e5531be8d0c.png)

After:
![image](https://user-images.githubusercontent.com/15073128/142918400-baa5b2c8-cc14-46ba-b02a-53c1c30f4368.png)

![image](https://user-images.githubusercontent.com/15073128/144300105-ecf116d9-ff37-45c2-b9b9-b110b43d172d.png)


### TESTING INSTRUCTIONS
Create a pivot table chart, click on download as csv icon. A dropdown with 2 options should open - Original and Pivoted.
Clicking Original should download a CSV file formatted like the one on the "Before" screenshot.
Clicking Pivoted should download a CSV file formatted like the one on the "After" screenshot

The CSV button on any other chart should work like before - begin download immediately instead of opening a dropdown menu.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

CC @junlincc

kgabryje committed:
feat(explore): export csv data pivoted for Pivot Table

kgabryje committed:
Implement dropdown with download csv options

kgabryje committed:
Change label to "Original"

kgabryje commented:
/testenv up

villebro commented on superset-frontend/src/explore/components/ExploreActionButtons.tsx:
```diff
@@ -133,6 +153,7 @@ const ExploreActionButtons = (props: ExploreActionButtonsProps) => {
     disabled: !canDownloadCSV,
   });
 
+  console.log({ latestQueryFormData });
```
oops!

kgabryje replied:
🤐

villebro commented on superset-frontend/src/explore/components/ExploreActionButtons.tsx:
```diff
@@ -42,18 +44,27 @@ type ExploreActionButtonsProps = {
   addDangerToast: Function;
 };
 
+const VIZ_TYPES_PIVOT = ['pivot_table', 'pivot_table_v2'];
```
nit: I wonder if we should call this `VIZ_TYPES_PIVOTABLE` to make it clear that it relates to whether or not the output can be pivoted or not, and not the chart type.

kgabryje replied:
Sounds good!

villebro commented on superset/charts/data/api.py:
```diff
@@ -221,6 +221,7 @@ def data(self) -> Response:
         if json_body is None:
             return self.response_400(message=_("Request is not JSON"))
 
+        form_data = json_body.pop("form_data", None)
```
Do we need to pop here? Can't we just do
```suggestion
        form_data = json_body.get("form_data")
```

kgabryje replied:
We don't need form_data in json_body. We pass it as a separate argument

kgabryje replied:
> > A few small comments, but this works and looks great! It would be nice if we could add a few tests:
> > 
> > 1. Ensure that a query context with `form_data` validates successfully
> > 2. Add tests for `ExportToCSVDropdown` that ensures it renders correctly with standard/pivoted options + the children.
> 
> Added tests for `ExportToCSVDropdown`. I didn't add tests for query context validation, as we pop form_data from json_body before it's used to create context, so there's no change there.

Edit: After some discussion, I implemented Ville's suggestions. We already had schema validations tests in place, so I just added `form_data` to query context generator.

villebro reviewed (commented):
A few small comments, but this works and looks great! It would be nice if we could add a few tests:
1. Ensure that a query context with `form_data` validates successfully
2. Add tests for `ExportToCSVDropdown` that ensures it renders correctly with standard/pivoted options + the children.

kgabryje committed:
Add tests

kgabryje commented:
> A few small comments, but this works and looks great! It would be nice if we could add a few tests:
> 
> 1. Ensure that a query context with `form_data` validates successfully
> 2. Add tests for `ExportToCSVDropdown` that ensures it renders correctly with standard/pivoted options + the children.

Added tests for `ExportToCSVDropdown`. I didn't add tests for query context validation, as we pop form_data from json_body before it's used to create context, so there's no change there.

kgabryje committed:
Add form data to query context

kgabryje committed:
Add form data to query context generator

kgabryje commented:
> > A few small comments, but this works and looks great! It would be nice if we could add a few tests:
> > 
> > 1. Ensure that a query context with `form_data` validates successfully
> > 2. Add tests for `ExportToCSVDropdown` that ensures it renders correctly with standard/pivoted options + the children.
> 
> Added tests for `ExportToCSVDropdown`. I didn't add tests for query context validation, as we pop form_data from json_body before it's used to create context, so there's no change there.

Edit: After some discussion, I implemented Ville's suggestions. We already had schema validations tests in place, so I just added form_data to query context generator.

villebro commented on tests/common/query_context_generator.py:
```diff
@@ -253,6 +253,7 @@ def generate(
                 )
             ],
             "result_type": ChartDataResultType.FULL,
+            "form_data": {},
```
`form_data` should probably be added to the signature of the method and passed as-is to here instead of defaulting to an empty dict.

villebro reviewed (commented):
A few last comments

kgabryje committed:
Explicitly make form_data optional

kgabryje commented on superset/charts/schemas.py:
```diff
@@ -1157,6 +1157,8 @@ class ChartDataQueryContextSchema(Schema):
     result_type = EnumField(ChartDataResultType, by_value=True)
     result_format = EnumField(ChartDataResultFormat, by_value=True)
 
+    form_data = fields.Raw()
```
Done

kgabryje commented on superset-frontend/packages/superset-ui-core/src/query/types/Query.ts:
```diff
@@ -158,6 +163,7 @@ export interface QueryContext {
   /** Response format */
   result_format: string;
   queries: QueryObject[];
+  form_data: QueryFormData;
```
Done

villebro reviewed (approved):
LGTM, thanks for this great improvement!
